src/Entity/SmsMessage.php

```diff
@@ -0,0 +1,411 @@
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getSenderEntity() {
+    return $this->get('sender_entity')->entity;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setSenderEntity(EntityInterface $entity) {
+    $this->set('sender_entity', $entity);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getRecipientEntity() {
+    return $this->get('recipient_entity')->entity;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setRecipientEntity(EntityInterface $entity) {
+    $this->set('recipient_entity', $entity);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getQueued() {
+    return $this->get('queued')->value;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setQueued($is_queued) {
+    $this->set('queued', $is_queued);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getCreatedTime() {
+    return $this->get('created')->value;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getSendTime() {
+    return $this->get('send_on')->value;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setSendTime($send_time) {
+    $this->set('send_on', $send_time);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getProcessedTime() {
+    return $this->get('processed')->value;
+  }
+  /**
+   * {@inheritdoc}
+   */
+  public function setProcessedTime($processed) {
+    $this->set('processed', $processed);
+    return $this;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function baseFieldDefinitions(EntityTypeInterface $entity_type) {
+    // Identifiers.
+    $fields['id'] = BaseFieldDefinition::create('integer')
+      ->setLabel(t('SMS message ID'))
+      ->setDescription(t('The SMS message ID.'))
+      ->setReadOnly(TRUE)
+      ->setSetting('unsigned', TRUE);
+
+    $fields['uuid'] = BaseFieldDefinition::create('uuid')
+      ->setLabel(t('UUID'))
+      ->setDescription(t('The SMS message UUID.'))
+      ->setReadOnly(TRUE);
+
+    $fields['gateway'] = BaseFieldDefinition::create('entity_reference')
+      ->setLabel(t('Gateway Plugin'))
+      ->setDescription(t('The gateway plugin instance.'))
+      ->setSetting('target_type', 'sms_gateway')
+      ->setReadOnly(TRUE)
+      ->setRequired(TRUE);
+
+    $fields['direction'] = BaseFieldDefinition::create('integer')
+      ->setLabel(t('Transmission direction'))
+      ->setDescription(t('Transmission direction, See SmsMessageInterface::DIRECTION_*.'))
+      ->setReadOnly(TRUE)
+      ->setSetting('unsigned', FALSE)
+      ->setSetting('size', 'tiny');
+
+    // Sender and receivers.
+    $fields['sender'] = BaseFieldDefinition::create('telephone')
+      ->setLabel(t('Sender phone number'))
+      ->setDescription(t('The phone number of the sender.'))
+      ->setDefaultValue('')
+      ->setRequired(FALSE);
+
+    $fields['sender_entity'] = BaseFieldDefinition::create('dynamic_entity_reference')
```

Comment:
Why do we need a separate `sender_entity` and `sender` field? We could just have a single `sender_entity` from whence we could have the number if such a field exists on the entity.